R/call.R

```diff
@@ -194,61 +183,117 @@ call_match <- function(call, fn, enum_dots) {
+#'   expression. If not supplied, the calling frame is used.
+#' @param .env Environment in which to look up call the function
+#'   definition and the contents of \code{...}. If not supplied, it is
+#'   retrieved from \code{call} if the latter is a frame object or a
+#'   formula.
+#' @param ...,.args Named expressions (constants, names or
+#'   calls) used to modify the call. Use \code{NULL} to remove
+#'   arguments.
+#' @seealso call_new
 #' @export
 #' @examples
 #' call <- quote(mean(x, na.rm = TRUE))
-#' call_standardise(call)
 #'
 #' # Modify an existing argument
-#' call_modify(call, list(na.rm = FALSE))
-#' call_modify(call, list(x = quote(y)))
+#' call_modify(call, na.rm = FALSE)
+#' call_modify(call, x = quote(y))
 #'
 #' # Remove an argument
-#' call_modify(call, list(na.rm = NULL))
+#' call_modify(call, na.rm = NULL)
 #'
 #' # Add a new argument
-#' call_modify(call, list(trim = 0.1))
+#' call_modify(call, trim = 0.1)
 #'
 #' # Add an explicit missing argument
-#' call_modify(call, list(na.rm = quote(expr = )))
+#' call_modify(call, na.rm = quote(expr = ))
+#'
+#' # Supply a list of new arguments with .args
+#' newargs <- list(na.rm = NULL, trim = 0.1)
+#' call_modify(call, .args = newargs)
 #'
 #' # If the call is missing, the parent frame is used instead.
-#' f <- function(bool = TRUE) call_modify(new_args = list(bool = FALSE))
+#' f <- function(bool = TRUE) call_modify(.args = list(bool = FALSE))
 #' f()
-call_modify <- function(call = NULL, new_args, env = NULL) {
-  stopifnot(is.list(new_args))
-  call <- call %||% call_frame(2)
-  call <- call_standardise(call, env)
+call_modify <- function(.call = NULL, ..., .args = list(), .env = NULL) {
+  stopifnot(is.list(.args))
+  args <- c(list(...), .args)
 
-  if (!all(has_names(new_args))) {
+  call <- .call %||% call_frame(2)
+  call <- call_standardise(call, .env)
+
+  if (!all(has_names(args))) {
     stop("All new arguments must be named", call. = FALSE)
   }
 
-  for (nm in names(new_args)) {
-    call[[nm]] <- new_args[[nm]]
+  for (nm in names(args)) {
+    call[[nm]] <- args[[nm]]
   }
   call
 }
 
-get_call <- function(call) {
+#' Extract function from a call
+#'
+#' @inheritParams call_standardise
+#' @export
+#' @seealso \code{\link{call_fn_name}}()
+#' @examples
+#' # Extract from a quoted call:
+#' call_fn(~matrix())
+#' call_fn(quote(matrix()))
+#'
+#' # Extract the calling function
+#' test <- function() call_fn()
+#' test()
+call_fn <- function(call = NULL, env = NULL) {
   if (is_frame(call)) {
-    call$expr
-  } else if (is_formula(call)) {
-    f_rhs(call)
+    call$fn
   } else {
-    call
+    info <- call_info(call, env, call_frame(2))
+    eval(info$call[[1]], info$env)
   }
 }
 
-#' Function name of a call
+#' Extract function name of a call
 #'
 #' @inheritParams call_standardise
-#' @return A string with the function name.
+#' @return A string with the function name, or \code{NULL} if the
```

Comment:
hmm I'm not sure. I think `NA`s make sense as an option type mostly in a vectorised context, and taking the name of a function feels like a scalar operation. E.g. I don't like testing with `is.na()` in a if branch because it is a vector function. Also `NA` has length 1 so you cannot use `length()` as a test like you can with `NULL`. Of course someone might want to use the function with `map_chr()`, in which case `NA` is better. It depends on the actual use-cases I guess.